src/Providers/Http/Collections/HeadersCollection.php

```diff
@@ -0,0 +1,144 @@
+     */
+    private array $headersMap = [];
+
+    /**
+     * Constructor.
+     *
+     * @since n.e.x.t
+     *
+     * @param array<string, string|list<string>> $headers Initial headers.
+     */
+    public function __construct(array $headers = [])
+    {
+        foreach ($headers as $name => $value) {
+            $this->set($name, $value);
+        }
+    }
+
+    /**
+     * Gets a specific header value.
+     *
+     * @since n.e.x.t
+     *
+     * @param string $name The header name (case-insensitive).
+     * @return list<string>|null The header value(s) or null if not found.
+     */
+    public function get(string $name): ?array
+    {
+        $lowerName = strtolower($name);
+        if (!isset($this->headersMap[$lowerName])) {
+            return null;
+        }
+
+        $actualName = $this->headersMap[$lowerName];
+        return $this->headers[$actualName];
+    }
+
+    /**
+     * Gets all headers.
+     *
+     * @since n.e.x.t
+     *
+     * @return array<string, list<string>> All headers with their original casing.
+     */
+    public function getAll(): array
+    {
+        return $this->headers;
+    }
+
+    /**
+     * Gets header values as a comma-separated string.
+     *
+     * @since n.e.x.t
+     *
+     * @param string $name The header name (case-insensitive).
+     * @return string|null The header values as a comma-separated string or null if not found.
+     */
+    public function getAsString(string $name): ?string
+    {
+        $values = $this->get($name);
+        return $values !== null ? implode(', ', $values) : null;
+    }
+
+    /**
+     * Checks if a header exists.
+     *
+     * @since n.e.x.t
+     *
+     * @param string $name The header name (case-insensitive).
+     * @return bool True if the header exists, false otherwise.
+     */
+    public function has(string $name): bool
+    {
+        return isset($this->headersMap[strtolower($name)]);
+    }
+
+    /**
+     * Sets a header value, replacing any existing value.
+     *
+     * @since n.e.x.t
+     *
+     * @param string $name The header name.
+     * @param string|list<string> $value The header value(s).
+     * @return void
+     */
+    private function set(string $name, $value): void
+    {
+        if (is_array($value)) {
+            $normalizedValues = array_values($value);
+        } else {
+            // Split comma-separated string into array
+            $normalizedValues = array_map('trim', explode(',', $value));
+        }
+        $lowerName = strtolower($name);
+
+        // If header exists with different casing, use the existing casing
+        if (isset($this->headersMap[$lowerName])) {
+            $actualName = $this->headersMap[$lowerName];
+            $this->headers[$actualName] = $normalizedValues;
+        } else {
+            // New header, use provided casing
+            $this->headers[$name] = $normalizedValues;
+            $this->headersMap[$lowerName] = $name;
+        }
```

Comment:
I would argue the opposite makes more sense: "If header exists with different casing, use the new casing"

The same way we set the new value here (obviously), I think we should also set the new name.
```suggestion
        // If header exists with different casing, replace it with provided casing.
        if (isset($this->headersMap[$lowerName])) {
            $actualName = $this->headersMap[$lowerName];
            if ($actualName !== $name) {
                unset($this->headers[$actualName]);
            }
        }

        $this->headers[$name] = $normalizedValues;
        $this->headersMap[$lowerName] = $name;
```

Reply:
Yeah, I agree! Resolved in 06f2da1